ethree-common/src/main/java/com/virgilsecurity/android/common/EThreeCore.kt

```diff
@@ -0,0 +1,771 @@
+import com.virgilsecurity.android.common.storage.local.GroupStorageFile
+import com.virgilsecurity.android.common.storage.local.KeyStorageLocal
+import com.virgilsecurity.android.common.storage.sql.SQLCardStorage
+import com.virgilsecurity.android.common.util.Const
+import com.virgilsecurity.android.common.util.Const.VIRGIL_BASE_URL
+import com.virgilsecurity.android.common.util.Const.VIRGIL_CARDS_SERVICE_PATH
+import com.virgilsecurity.android.common.worker.*
+import com.virgilsecurity.common.model.Completable
+import com.virgilsecurity.common.model.Data
+import com.virgilsecurity.common.model.Result
+import com.virgilsecurity.keyknox.build.VersionVirgilAgent
+import com.virgilsecurity.sdk.cards.Card
+import com.virgilsecurity.sdk.cards.CardManager
+import com.virgilsecurity.sdk.cards.validation.VirgilCardVerifier
+import com.virgilsecurity.sdk.client.HttpClient
+import com.virgilsecurity.sdk.client.VirgilCardClient
+import com.virgilsecurity.sdk.crypto.*
+import com.virgilsecurity.sdk.jwt.Jwt
+import com.virgilsecurity.sdk.jwt.accessProviders.CachingJwtProvider
+import com.virgilsecurity.sdk.jwt.contract.AccessTokenProvider
+import com.virgilsecurity.sdk.storage.DefaultKeyStorage
+import com.virgilsecurity.sdk.storage.KeyStorage
+import java.io.InputStream
+import java.io.OutputStream
+import java.util.*
+
+/**
+ * [EThreeCore] class simplifies work with Virgil Services to easily implement End to End Encrypted
+ * communication.
+ */
+abstract class EThreeCore
+/**
+ * @constructor Initializing [CardManager] with provided in [EThreeCore.initialize] callback
+ * [onGetTokenCallback] using [CachingJwtProvider] also initializing [DefaultKeyStorage] with
+ * default settings.
+ */
+constructor(identity: String,
+            getTokenCallback: OnGetTokenCallback,
+            keyChangedCallback: OnKeyChangedCallback?,
+            context: Context) {
+
+    private var accessTokenProvider: AccessTokenProvider
+    private val rootPath: String
+
+    private val cloudKeyManager: CloudKeyManager
+    val lookupManager: LookupManager
+    lateinit var keyStorageLocal: KeyStorageLocal
+        private set
+    private lateinit var authorizationWorker: AuthorizationWorker
+    private lateinit var backupWorker: BackupWorker
+    private lateinit var groupWorker: GroupWorker
+    private lateinit var p2pWorker: PeerToPeerWorker
+    private lateinit var searchWorker: SearchWorker
+
+    var groupManager: GroupManager? = null
+        private set
+
+    protected val crypto: VirgilCrypto = VirgilCrypto()
+
+    protected abstract val keyStorage: KeyStorage
+
+    val cardManager: CardManager
+    val identity: String
+
+    init {
+        this.identity = identity
+        val cardCrypto = VirgilCardCrypto(crypto)
+        val virgilCardVerifier = VirgilCardVerifier(cardCrypto)
+        val httpClient = HttpClient(Const.ETHREE_NAME, VersionVirgilAgent.VERSION) // FIXME wrong VersionVirgilAgent - from keyknox
```

Comment:
Not resolved fixme. Can affect statistics, I guess.

Reply:
done